### DISCUSSION TITLE
[Image secret scan]: Memory consumption too high

### DISCUSSION THREAD

**Discussion by rajaSahil** [aquasecurity/trivy#7078] [Bugs]
### Description

When I try to scan image for vulnerabilities/secrets, it starts consuming almost 7GB of memory and due to which pods are getting into `OOMKilled` state. 

**Image**:  ` rajasahil/test-large-image:v1`

**Command**:
`trivy image rajasahil/test-large-image:v1`

### Questions:
-  Is there a way through which we can let trivy know to use less resources and do the scan?
-  Do we know how much maximum resources can trivy use? It would be helpful to specify resources limits for a pod and would not see issues such as `OOMKilled`

### Desired Behavior

- There should be a way through which we can let trivy know to use less resources and do the scan.
  - `slow` flag should be available for image scan too.
  - upper limits of resources can be specified to trivy which will not allow trivy to consume more resources than specified limit.

- There should be a benchmark for trivy which should speicify how much max resource trivy would utilize. It would be helpful to specify resources limits for a pod and would not see issues such as `OOMKilled` 

### Actual Behavior

Due to huge resource utilization, pods getting `OOMKilled`

### Issue
**As soon as scan starts, Trivy ends up utilizng 7GB of memory.**

### Reproduction Steps

```bash
1. Run command `trivy image rajasahil/test-large-image:v1`
```

### Target

Container Image

### Scanner

Vulnerability,Secret

### Output Format

JSON

### Mode

Standalone

### Debug Output

```bash
Unable to get output, since trivy ends up hanging my laptop and pods are also getting killed.
```

### Operating System

Linux, Mac os

### Version

```bash
Version: 0.53.0
Vulnerability DB:
  Version: 2
  UpdatedAt: 2024-07-02 06:11:21.992759055 +0000 UTC
  NextUpdate: 2024-07-02 12:11:21.992758684 +0000 UTC
  DownloadedAt: 2024-07-02 06:18:45.926729538 +0000 UTC
```

### Checklist

- [X] Run `trivy clean --all`
- [X] Read [the troubleshooting](https://aquasecurity.github.io/trivy/latest/docs/references/troubleshooting/)

**Comment by knqyf263** [comment]
It works in my environment.

```
$ trivy image rajasahil/test-large-image:v1 --scanners vuln 
```

**Reply by knqyf263** [reply]
It consumes 130-140MB.

```
$ time -f "Memory:%M KB time:%E" ./trivy image rajasahil/test-large-image:v1 --scanners vuln
2024-07-02T13:04:34+04:00       INFO    Vulnerability scanning is enabled
2024-07-02T13:05:28+04:00       INFO    Number of language-specific files       num=0
Memory:135472 KB time:0:54.85
```

**Reply by rajaSahil** [reply]
Remove vuln, I'm not specifying any scanners. So it would scan all scans i.e vulnerabilities, secrets, etc.

Command should be
`time -f "Memory:%M KB time:%E" ./trivy image rajasahil/test-large-image:v1 `

**Reply by knqyf263** [reply]
You want to scan vulns, no?

**Reply by rajaSahil** [reply]
Yes, but it should also scan secrets. By default the behaviour is to scan for vulnerabilities and secrets, that's what I want.

**Comment by rajaSahil** [comment]
To be more precise, issue is happening with respect to secret scanning.

`trivy image rajasahil/test-large-image:v1 --scanners secret`

**Reply by knqyf263** [reply]
As displayed in log messages, secret scanning is not memory efficient by design. I guess your image has a large file. @DmitriyLewen Could you take a look? We might want to show warnings about large files so users can skip them.

**Reply by DmitriyLewen** [reply]
you are right. 
`rajasahil/test-large-image:v1` contains one big file:
```bash
/ # ls -hl
total 946M   
...
-rwxr-xr-x    1 root     root      946.4M Jun 30 02:42 fslibs-entry-point.sh
...
```
I am adding warning.

**Reply by rajaSahil** [reply]
@DmitriyLewen Is there a way to let trivy know if large file exists, skip secret scanning?

**Reply by rajaSahil** [reply]
This can be default behaviour for trivy image scanning

**Reply by DmitriyLewen** [reply]
I created #7085 to display a warning with the file path and size.

**Comment by itaysk** [comment]
related: https://github.com/aquasecurity/trivy/discussions/7151
